src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs

```diff
@@ -2904,116 +2867,123 @@ private static bool IsReallyAType(TypeSymbol type)
                 (type.SpecialType != SpecialType.System_Void);
         }
 
-        private struct Candidates
+        private static void GetAllCandidates(Dictionary<TypeSymbolWithAnnotations, TypeSymbolWithAnnotations> candidates, ArrayBuilder<TypeSymbolWithAnnotations> builder)
         {
-            // The keys are types, compared ignoring dynamic, tuple names, and nullability.
-            // The values are different nullabilities associated with that type. Both keys and
-            // values are TypeSymbolWithAnnotations, although we could use bool[]? for values
-            // and recreate the TypeSymbolWithAnnotations from the key+value as needed.
-            private readonly Dictionary<TypeSymbolWithAnnotations, List<TypeSymbolWithAnnotations>> _dictionary;
-            private readonly EqualsIgnoringDynamicAndTupleNamesComparer _valueComparer;
-
-            internal Candidates(bool includeNullability)
-            {
-                _dictionary = new Dictionary<TypeSymbolWithAnnotations, List<TypeSymbolWithAnnotations>>(
-                    EqualsIgnoringDynamicAndTupleNamesComparer.WithoutNullability);
-                _valueComparer = includeNullability ?
-                    EqualsIgnoringDynamicAndTupleNamesComparer.WithNullability :
-                    EqualsIgnoringDynamicAndTupleNamesComparer.WithoutNullability;
-            }
+            builder.AddRange(candidates.Values);
+        }
 
-            internal void GetAll(ArrayBuilder<TypeSymbolWithAnnotations> builder)
+        private static bool AddAllCandidates(
+            Dictionary<TypeSymbolWithAnnotations, TypeSymbolWithAnnotations> candidates,
+            HashSet<TypeSymbolWithAnnotations> bounds,
+            VarianceKind variance,
+            ConversionsBase conversions)
+        {
+            foreach (var candidate in bounds)
             {
-                foreach (var values in _dictionary.Values)
+                var type = candidate;
+                if (!conversions.IncludeNullability)
                 {
-                    builder.AddRange(values);
+                    type = type.SetUnknownNullabilityForReferenceTypes();
```

Comment:
>type = type.SetUnknownNullabilityForReferenceTypes(); [](start = 20, length = 53)

Do we need to change the nullability of candidates we put in the dictionary? Or might it be sufficient to use a proper comparer (ie. that ignores nullability in some scenarios)?

Reply:
Did you have a thought on this?

---
In reply to: [223843555](https://github.com/dotnet/roslyn/pull/29727#discussion_r223843555) [](ancestors = 223843555)